cmd/nephe-controller/secret_webhook.go

```diff
@@ -0,0 +1,169 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/go-logr/logr"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	controllerclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	crdv1alpha1 "antrea.io/nephe/apis/crd/v1alpha1"
+)
+
+// SecretValidator is used to validate Secret
+type SecretValidator struct {
+	Client  controllerclient.Client
+	Log     logr.Logger
+	decoder *admission.Decoder
+}
+
+// Handle handles admission requests for Secret.
+func (v *SecretValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	v.Log.Info("Received webhook", "req", req, "operation", req.Operation)
+	switch req.Operation {
+	case admissionv1.Create:
+		return v.validateCreate(req)
+	case admissionv1.Update:
+		return v.validateUpdate(req)
+	case admissionv1.Delete:
+		return v.validateDelete(req)
+	default:
+		v.Log.Info("Invalid admission webhook request")
+	}
+	return admission.Allowed("")
+}
+
+// InjectDecoder injects the decoder.
+func (v *SecretValidator) InjectDecoder(d *admission.Decoder) error { //nolint:unparam
+	v.decoder = d
+	return nil
+}
+
+// allowSecretUpdate returns true only when Secret data key is unchanged
+func (v *SecretValidator) allowSecretUpdate(new *corev1.Secret, old *corev1.Secret, key string) bool {
+	if changed := string(new.Data[key]) != string(old.Data[key]); changed {
+		return false
+	}
+	return true
+}
+
+// getCPABySecret returns nil only when the Secret is not used by
+// CloudProvideAccount,
+// Otherwise, an appropriate error message and the corresponding
+// CloudProvideAccount CR will be returned.
+func (v *SecretValidator) getCPABySecret(newSecret *corev1.Secret, oldSecret *corev1.Secret,
+	isDelete bool) (error, *crdv1alpha1.CloudProviderAccount) {
+	// TODO: Figure out how to get all CRs
+	cpaList := &crdv1alpha1.CloudProviderAccountList{}
+	options := &controllerclient.ListOptions{Limit: 1000}
+	err := v.Client.List(context.TODO(), cpaList, options)
+	if err != nil {
+		return fmt.Errorf("failed to get CloudProviderAccount list, err:%v", err), nil
+	}
+
+	for _, cpa := range cpaList.Items {
+		v.Log.Info("Checking CloudProvideAccount", "Name", cpa.Name, "Namespace", cpa.Namespace)
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AWSConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					v.Log.Info("Cannot delete", "Secret", oldSecret.Name, "CloudProvideAccount", cpa.Name)
```

Comment:
update this logging, else error message and this log is the same.. I believe you want to highlight its used by CPA

Reply:
I just want to have a log message which captures an attempt was made to delete/update secret which is used by CPA.
Sure, I will change it to ERROR message.

Reply:
Changing the logs type to ERROR, will dump the call-stack. Also this is not an error, so IMO it should be an INFO message only.
 
2022-09-07T08:53:46.430Z        ERROR   webhook.Secret  webhook/secret_webhook.go:56    Secret resource cannot be deleted       {"Secret": "aws-account-creds", "CloudProviderAccount": "cloudprovideraccount-aws-sample"}
antrea.io/nephe/pkg/apiserver/webhook.(*SecretValidator).Handle
        /usr/src/antrea.io/nephe/pkg/apiserver/webhook/secret_webhook.go:56
sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).Handle
        /tmp/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/webhook/admission/webhook.go:146
sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).ServeHTTP
        /tmp/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/webhook/admission/http.go:98
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1
        /tmp/gopath/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:40
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1
        /tmp/gopath/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:117
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2
        /tmp/gopath/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:84
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2109
net/http.(*ServeMux).ServeHTTP
        /usr/local/go/src/net/http/server.go:2487
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2947
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1991